feat(spl): scripts and GitHub workflow support for running spl-token proofs

Stevengre opened the pull request:
- setup the scripts to run spl proofs;
- modify GitHub workflow configuration to run spl proofs remotely.

Stevengre committed:
refactor(sync-spl-specs): enhance output configuration and function overrides for better clarity and maintainability

Stevengre committed:
feat(submodule): add program/test-properties/mir-semantics submodule for enhanced testing

Stevengre committed:
feat(setup): add setup script for spl-token property tests with kmir

Stevengre commented:
Here is an error when running `./setup.sh` in `program/test-properties/`:
```
thread 'rustc' panicked at src/printer.rs:789:13:
assertion failed: kind.is_fn_ptr()
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: could not compile `autocfg` (lib)

Caused by:
  process didn't exit successfully: `/home/zhaoji/solana-token/program/test-properties/mir-semantics/deps/.stable-mir-json/release.sh --crate-name autocfg --edition=2015 /home/zhaoji/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/autocfg-1.4.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=165 --crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values())' -C metadata=6ee610fba99ef989 -C extra-filename=-3efdef99bf5440dc --out-dir /home/zhaoji/solana-token/target/debug/deps -L dependency=/home/zhaoji/solana-token/target/debug/deps --cap-lints allow` (exit status: 101)
warning: build failed, waiting for other jobs to finish...
Items missing static with id StaticDef(DefId { id: 638, name: "core::unicode::unicode_data::white_space::WHITESPACE_MAP" }) and name "_ZN4core7unicode12unicode_data11white_space14WHITESPACE_MAP17hc3c35dc72f9ae8e8E"
```

jberthold commented:
> Here is an error when running `./setup.sh` in `program/test-properties/`:
> ...
> Items missing static with id StaticDef(DefId { id: 638, name:... 

It comes from this code: https://github.com/runtimeverification/stable-mir-json/blob/b54152a3099c4756c31fea1a126c20167ae42798/src/printer.rs#L1303-L1317

Stevengre committed:
feat(tests): add .gitignore and proofs documentation for test properties

Stevengre committed:
feat: support spl proof workflow

Stevengre committed:
refactor(workflow): reduce vars

Stevengre committed:
refactor(workflow): keep consistency

jberthold commented on program/test-properties/mir-semantics:
So this basically duplicates the entire proof-running setup from `p-token/test-properties` to here, and we are using `mir-semantics` _twice_ as a submodule.
I would much prefer if we could merge the two uses into one submodule. The different cheatcodes in `mir-semantics` should be OK to exist side by side because we are targeting different functions. I don't know if any code can be shared between them but duplicating the entire `mir-semantics` is not required even if nothing is shared.

Stevengre replied:
deleted the duplicate submodule

jberthold commented on program/src/entrypoint-runtime-verification.rs:
```diff
@@ -190,17 +189,15 @@ fn inner_process_instruction(
+/// accounts[1] // Destination Info
+/// accounts[2] // Authority Info
+/// accounts[3..14] // Signers
+/// instruction_data[0..8] // Little Endian Bytes of u64 amount
```
This is probably not correct because the first byte should still be the discriminator, or not?
In line 206 below, `cheatcode_set_discriminator` is applied, and in 210-211 the "real" instruction data is extracted as bytes [1..9].

Stevengre replied:
I don't provide a rule to transform the comments in the p-token one. Therefore, it remains the same as it is in the p-token one.

Stevengre replied:
- Solved with 4752189c5103e02a58435007379b6c24dc727788

jberthold commented on program/src/entrypoint-runtime-verification.rs:
```diff
@@ -247,150 +244,175 @@ fn test_process_transfer(
-
-                        if unsigned_exists {
-                            assert_eq!(result, Err(ProgramError::MissingRequiredSignature));
+            { // Validate Owner
```
This looks like a double `{` to me... but OK to leave if it is an artefact of the automatic conversion

Stevengre replied:
This is the same as what it is in the p-token:

https://github.com/runtimeverification/solana-token/blob/d2abef88aba91e78d2eb4e527e0b9332fc6bffb7/p-token/src/entrypoint-runtime-verification.rs#L641

jberthold commented on program/src/entrypoint-runtime-verification.rs:
```diff
@@ -247,150 +244,175 @@ fn test_process_transfer(
+                    assert_eq!(result, Err(ProgramError::MissingRequiredSignature));
+                    return result;
+                }
             }
```
The extra curly braces from above are closed here.

jberthold commented on program/src/entrypoint-runtime-verification.rs:
```diff
@@ -247,150 +244,175 @@ fn test_process_transfer(
-                assert_eq!(result, Err(ProgramError::MissingRequiredSignature));
-                return result;
+                // Line 133-135: Non-multisig case - check if owner_account_info.is_signer()
+                else if !accounts[2].is_signer {
+                    assert_eq!(result, Err(ProgramError::MissingRequiredSignature));
+                    return result;
+                }
             }
 
             if old_src_delgated_amount < amount {
                 assert_eq!(result, Err(ProgramError::Custom(1)));
                 return result;
             }
         } else {
-            // Validate Owner for non-delegate case
-            if src_owner != *accounts[2].key { // TODO Daniel: CHANGE deref keys
-                assert_eq!(result, Err(ProgramError::Custom(4)));
-                return result;
-            } else if accounts[2].data_len() == Multisig::LEN && accounts[2].owner == &crate::id() {
-                // #[cfg(feature="multisig")]
-                {
-                    if multisig_is_initialised.is_err() {
-                        assert_eq!(result, Err(ProgramError::InvalidAccountData));
-                        return result;
-                    } else if !multisig_is_initialised.unwrap() {
-                        assert_eq!(result, Err(ProgramError::UninitializedAccount));
-                        return result;
-                    } else {
-                        let multisig = get_multisig(&accounts[2]);
-
-                        let unsigned_exists = accounts[3..].iter()
-                            .any(|potential_signer| {
-                                multisig.signers() // TODO Daniel: CHANGE .signers to .signers()
-                                    .iter()
-                                    .any(|registered_key| registered_key == potential_signer.key && !potential_signer.is_signer)
-                            });
-
-                        if unsigned_exists {
-                            assert_eq!(result, Err(ProgramError::MissingRequiredSignature));
+            { // Validate Owner
```
Same double `{`...

Stevengre replied:
Same as p-token here

https://github.com/runtimeverification/solana-token/blob/d2abef88aba91e78d2eb4e527e0b9332fc6bffb7/p-token/src/entrypoint-runtime-verification.rs#L701

The script only transforms the original code without constructing the blocks by itself.

jberthold commented on .github/workflows/rv-run-proofs.yaml:
```diff
@@ -123,32 +159,33 @@ jobs:
+
           timeout ${{ inputs.timeout }} \
             docker exec \
               --workdir /workdir "${KMIR_CONTAINER_NAME}" \
-              kmir prove-rs --smir artefacts/p-token.smir.json \
-              --start-symbol 'pinocchio_token_program::entrypoint::${{ matrix.proof }}' \
+              kmir prove-rs --smir "artefacts/${{ needs.compile.outputs.artifact_name }}.json" \
+              --start-symbol "${{ needs.compile.outputs.start_prefix }}${{ matrix.proof }}" \
               --verbose \
               --max-depth 500 \
               --max-iterations 100 \
               --proof-dir artefacts/proof \
             || echo "Runner signals proof failure"
```
The proofs are run in a docker container based on branch `feature/p-token` (line 119), so this is one main reason why we want both P-token and SPL-token cheatcodes to be in the same `mir-semantics` branch.
We can of course add another parameter to this workflow to select the suitable `mir-semantics` docker image but let's try to have both in the same branch .

Stevengre replied:
Understood! Removed the mir-semantics under program folder.

jberthold commented on program/test-properties/run-proofs.sh:
Did you have to make any changes to this script, compared to the `p-token/test-properties/run-proofs.sh` script?
If not, we can use a symlink instead of copying the file, or even call a shared script (which is of course more work, so better use a symlink).
If any changes were made, what were they?

Stevengre replied:
Before, since some of the configs are not the same, I provided separate files to keep the original one unchanged. But it does introduce many duplicates. 

Currently, I tried to provide minimal changes to the p-token scripts and provide links in the spl-token to call the same script with different environments.

Stevengre committed:
remove duplicate mir-semantics

Stevengre committed:
refactor: p-token scripts for spl-token

Stevengre committed:
fix: ./setup.sh

Stevengre committed:
fix: spl-token test's docstrings

jberthold reviewed (approved):
Thanks for modifying the `run-proofs.sh`. 
The changes in the workflow and scripts LGTM

Stevengre committed:
feat(spl): generate all harness functions from p-token to spl-token (#60)

- Transform all `p-token` specs to `spl-token` with
`./program/test-properties/scripts/sync-spl-specs.py`.
- Pass `pnpm programs:build -- --features rvo && pnpm programs:test --
--features rvo` by commenting out `6 - Set Authority Account` and `12 -
Transfer Checked`
